packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -1,242 +1,265 @@
+        }
+      },
+    );
+
+    onPreferencesStateChange(
+      ({ selectedAddress: newSelectedAddress, useTokenDetection }) => {
+        const isSelectedAddressChanged =
+          this.#selectedAddress !== newSelectedAddress;
+        const isDetectionChangedFromPreferences =
+          this.#isDetectionEnabledFromPreferences !== useTokenDetection;
+
+        this.#selectedAddress = newSelectedAddress;
+        this.#isDetectionEnabledFromPreferences = useTokenDetection;
+
+        if (
+          useTokenDetection &&
+          (isSelectedAddressChanged || isDetectionChangedFromPreferences)
+        ) {
+          this.detectTokens();
+        }
+      },
+    );
+
+    this.messagingSystem.subscribe(
+      'NetworkController:stateChange',
+      ({ providerConfig: { chainId: newChainId } }) => {
+        this.#isDetectionEnabledForNetwork =
+          isTokenDetectionSupportedForNetwork(newChainId);
+        const isChainIdChanged = this.#chainId !== newChainId;
+
+        this.#chainId = newChainId;
+
+        if (this.#isDetectionEnabledForNetwork && isChainIdChanged) {
+          this.detectTokens();
+        }
+      },
+    );
+
+    this.#isDetectionEnabledFromPreferences = defaultUseTokenDetection;
+    this.#isDetectionEnabledForNetwork = isTokenDetectionSupportedForNetwork(
+      this.#chainId,
+    );
   }
 
   /**
    * Start polling for detected tokens.
    */
   async start() {
-    this.configure({ disabled: false });
-    await this.startPolling();
+    this.#disabled = false;
+    await this.#startPolling();
   }
 
   /**
    * Stop polling for detected tokens.
    */
   stop() {
-    this.configure({ disabled: true });
-    this.stopPolling();
+    this.#disabled = true;
+    this.#stopPolling();
   }
 
-  private stopPolling() {
-    if (this.intervalId) {
-      clearInterval(this.intervalId);
+  #stopPolling() {
+    if (this.#intervalId) {
+      clearInterval(this.#intervalId);
     }
   }
 
   /**
    * Starts a new polling interval.
-   *
-   * @param interval - An interval on which to poll.
    */
-  private async startPolling(interval?: number): Promise<void> {
-    interval && this.configure({ interval }, false, false);
-    this.stopPolling();
+  async #startPolling(): Promise<void> {
+    this.#stopPolling();
     await this.detectTokens();
-    this.intervalId = setInterval(async () => {
-      await this.detectTokens();
-    }, this.config.interval);
+    this.#intervalId = setInterval(() => {
+      this.detectTokens();
+    }, this.getIntervalLength());
   }
 
-  private getCorrectChainId(networkClientId?: NetworkClientId) {
+  #getCorrectChainId(networkClientId?: NetworkClientId) {
     if (networkClientId) {
-      return this.getNetworkClientById(networkClientId).configuration.chainId;
+      const {
+        configuration: { chainId },
+      } = this.messagingSystem.call(
+        'NetworkController:getNetworkClientById',
+        networkClientId,
+      );
+      this.#chainId = chainId;
```

Comment:
We now have a `getNetworkConfigurationByNetworkClientId` action that we can use as more of a straight shot:

``` suggestion
      const networkConfiguration = this.messagingSystem.call(
        'NetworkController:getNetworkConfigurationByNetworkClientId',
        networkClientId,
      );
      this.#chainId = networkConfiguration.chainId;

Reply:
Fixed: https://github.com/MetaMask/core/commit/9c8491a0a98c512999b4f789b50c358c7e061df2

Reply:
Reverted this here: https://github.com/MetaMask/core/pull/3609/commits/173168f172ec1b1b56139c6a7ecb559c49ad7c36

Requires too many changes in tests. Refactors to use this action will need to happen after all tests are passing.

Reply:
Also, this suggestion below uses the `getNetworkClientById`: https://github.com/MetaMask/core/pull/3609#discussion_r1428323306